synapse/rest/client/v1/logout.py

```diff
@@ -74,6 +78,38 @@ def on_OPTIONS(self, request):
         return 200, {}
 
 
+class SAMLLogoutServlet(LogoutRestServlet):
+    def __init__(self, hs):
+        super().__init__(hs)
+        self._saml_handler = hs.get_saml_handler()
+
+    async def on_POST(self, request):
+        requester = await self.auth.get_user_by_req(request)
+        # Logout from the synapse server
+        await self._logout(requester, request)
+        # Try to use the SAML logout endpoint.
+        # It may fail if the user logged in via m.login.password
+        # TODO: find a way to know is the user logged in via
+        # m.login.password or via m.login.sso / .token
+        logout_url = self._saml_handler.create_logout_request(
+            requester.user.to_string(),
+            self.auth.get_access_token_from_request(request),
+        )
+        if logout_url:
+            set_cors_headers(request)
+            request.redirect(logout_url)
+            finish_request(request)
+            # Returning None because the response is handle by _saml_handler.logout
+            # instead of the synapse.http.HttpServer interface with JsonResource.
+            # See server.py for more information.
+            return None
+
+        return 200, {}
+
+
 def register_servlets(hs, http_server):
-    LogoutRestServlet(hs).register(http_server)
+    if hs.config.saml2_enabled:
+        SAMLLogoutServlet(hs).register(http_server)
+    else:
+        LogoutRestServlet(hs).register(http_server)
```

Comment:
I'm not super-keen on having different servlet implementations bound to the same path. Can you just update the existing `LogoutRestServlet` to do the saml stuff if saml is enabled?

Reply:
It's almost the same implementation : SAMLLogoutServlet inherits LogoutRestServlet. The idea was to have no SAML logic inside LogoutRestServlet so it's cleaner. Inside SAMLLogoutServlet we re-use the whole logic with a call to self._logout() that comes from LogoutRestServlet.
At first I implemented all inside LogoutRestServlet as suggested but it doesn't fill right to mix standard logout with SAML especially if in the future we add on OpenID. Also the login servlet tried to separate that logic as well but it's easier for the login because there is a different path for sso (/redirect).
What do you think after my explanation ?

Reply:
I'm afraid I still think that having a single path which can be implemented by two different servlets is inconsistent with the way everything else works in synapse. It's fine if the servlet immediately defers to one of two implementations; I just think it's hard to follow the code the way you've written it.

> Also SAMLLogoutServlet inherits LogoutRestServlet and re-uses the constructor and _logout() to share some implementations.

Yeah. I'm not a big fan of this, either. It's easy to write confusing code with inheritance.